weave/tests/test_weave_client.py

```diff
@@ -1165,3 +1166,167 @@ def test_table_partitioning(network_proxy_client):
         + 1  # the second  create
         + NUM_ROWS / 2  # updates - 2 per batch
     )
+
+
+def test_summary_tokens_cost(client):
+    is_sqlite = isinstance(client.server._internal_trace_server, SqliteTraceServer)
+    if is_sqlite:
+        # SQLite does not support costs
+        return
+
+    @weave.op()
+    def gpt4(text):
+        result = "a: " + text
+        return {
+            "result": result,
+            "model": "gpt-4",
+            "usage": {
+                "prompt_tokens": 1000000,
+                "completion_tokens": 2000000,
+            },
+        }
+
+    @weave.op()
+    def gpt4o(text):
+        result = "bbbb: " + text
+        return {
+            "result": result,
+            "model": "gpt-4o",
+            "usage": {
+                "prompt_tokens": 3000000,
+                "completion_tokens": 5000000,
+            },
+        }
+
+    @weave.op()
+    def models(text):
+        return (
+            gpt4(text)["result"]
+            + " "
+            + gpt4(text)["result"]
+            + " "
+            + gpt4o(text)["result"]
+        )
+
+    res = models("hello")
+    assert res == "a: hello a: hello bbbb: hello"
+
+    call = list(models.calls())[0]
+
+    assert call.summary["usage"] == {
+        "gpt-4": {
+            "requests": 2,
+            "prompt_tokens": 2000000,
+            "completion_tokens": 4000000,
+        },
+        "gpt-4o": {
+            "requests": 1,
+            "prompt_tokens": 3000000,
+            "completion_tokens": 5000000,
+        },
+    }
+
+    callsWithCost = list(
+        client.calls(
+            weave_client._CallsFilter(op_names=[call.op_name]),
+            include_costs=True,
+        )
+    )
+    callsNoCost = list(
+        client.calls(
+            weave_client._CallsFilter(op_names=[call.op_name]),
+            include_costs=False,
+        )
+    )
+
+    assert len(callsWithCost) == len(callsNoCost)
+    assert len(callsWithCost) == 1
+
+    noCostCallSummary = callsNoCost[0].summary
+    withCostCallSummary = callsWithCost[0].summary
+
+    assert withCostCallSummary.get("weave", "bah") != "bah"
+    assert len(withCostCallSummary["weave"]["costs"]) == 2
+
+    gpt4cost = withCostCallSummary["weave"]["costs"]["gpt-4"]
+    gpt4ocost = withCostCallSummary["weave"]["costs"]["gpt-4o"]
+
+    # rounds to nearest integer avoids .999999992345234
```

Comment:
tip: use `pytest.approx` - it is used in various parts of the code base for reference.